doc/design/audit-trail-mvp.md

```diff
@@ -0,0 +1,152 @@
+future.
+
+For context: because most changes to an `Asset` model cause a new model instance
+to be created, and because that new `Asset` instance points to its "ancestor"
+via a `previous` property, the history of an asset is at least partially
+traceable through this "asset chain".
+
+### Audit
+
+| Action | Recoverable fields | Unrecoverable fields | Notes |
+| --- | --- | --- | --- |
+| Create Dandiset | date_time, dandiset_id | user_id, metadata | date_time is equal to the Dandiset’s created_at value. user_id and metadata (as they were at the beginning of the Dandiset’s lifecycle) cannot be known with certainty.<br><br> Dandisets that have already been deleted will not have any recoverable fields (though we believe it very unlikely that a Dandiset has ever actually been deleted).<br><br> Independent of asset chains. |
+| Add owner | none | date_time, user_id, dandiset_id, owner_id | These events are all "summed" into the current owners list, with no way to know the evolution thereof.<br><br> Independent of asset chains. |
+| Remove owner | none | date_time, user_id, dandiset_id, owner_id | These events are all "summed" into the current owners list, with no way to know the evolution thereof.<br><br> Independent of asset chains. |
+| Update Dandiset metadata | none | date_time, user_id, dandiset_id, metadata | These events are all "summed" into the current Dandiset metadata, with no way to know the evolution thereof.<br><br> Independent of asset chains. |
+| Add asset | date_time, dandiset_id, asset_path, checksum, metadata | user_id | The last Asset instance in the chain (i.e., the one that has a null previous pointer) represents the version of the asset that was first added to a Dandiset. Everything about this asset besides the user who uploaded it is available in the Asset instance, the AssetBlob it points to, and the association table between Dandisets and Assets that relates them.<br><br> The dandiset_id may not be recoverable for Asset instances that have been removed from a Dandiset, or for Dandisets that were deleted. |
+| Update asset | date_time, dandiset_id, asset_path, checksum, metadata | user_id | As above, the user_id information is missing, and the dandiset_id may be as well. |
+| Remove asset | asset_path | date_time, user_id, dandiset_id | A removed asset will be one at the head of an asset chain that has no association to a Dandiset. Unfortunately, without that association, the dandiset_id cannot be recovered, and because it is the absence of a relation that indicates removal, the date_time value is also unrecoverable. |
+| Publish Dandiset | date_time, dandiset_id, version | user_id | The date_time, dandiset_id, and version are all available as part of the Version instance’s metadata. |
+| Delete Dandiset | none | date_time, user_id, dandiset_id | Deleted Dandisets just don’t exist anymore, so no information (even the dandiset_id) is available. |
+
+Note that the majority of information describing these auditable actions is
```

Comment:
Is there a reason that an audit log could not be preserved and provided for deleted dandisets?

Deletion seems important to know which user may have deleted the set when (assumingely downstream users of a dandiset would want to know if it no longer exists as well -- especially for scientific reproducibility)

Curious to know your thoughts here -- if there is a way of noting historical evidence of a given dandiset, even though the dandiset's data is no longer available.

Reply:
My comments in this part of the doc are about recovering _historical_ events from the existing record. Indeed, it is possible, valuable, and in our plans to record deletion events and keep them in the audit record for a given Dandiset. That is, the Dandiset may be deleted, but we'll still have the full history of that Dandiset, even though the data itself is gone.

It may be worth noting that deleting Dandisets is not a generally available action: it is only possible for Dandisets that have not published any versions (with a major exception for emergency situations, such as a lab accidentally publishing something they should not have, etc.; in these situations we intervene to solve the problem out-of-band).